---
title: "eDNA analysis"
author: "Stefan Filges"
date: '`r format(Sys.Date())`'
output: 
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
    collapsed: true
    smooth_scroll: true
    number_sections: true
    theme: spacelab
    highlight: tango
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  cache = TRUE,
  message = FALSE,
  fig.align = 'center',
  warning = FALSE
)
```

# Analysis

```{r library-import, message=FALSE, warning=FALSE}
library(tidyverse)
library(dada2)
library(phyloseq)
library(Biostrings)
```

```{r data-import}
# Consensus reads with UMI family sizes
cons <- readr::read_csv("out.csv")
print(head(cons))

# Fasta with reference sequences for taxonomy assignment
# This reference fasta file should be formatted so that the
# id lines correspond to the taxonomy
taxo_db <- "../data/scandi_fish_12s_v1.4.fa"
```

## Assigning taxa

```{r assign_taxa}
# Convert cons table to dataframe in dada2 input format
seqtab_nochim <- data.frame(
  sequence = cons["Read"],
  abundance = cons["Size"]
)
colnames(seqtab_nochim) <- c("sequence", "abundance")

# This uses a Naives-Bayes trained classifier
# seqs: A character vector of the sequences to be assigned
taxa <- dada2::assignTaxonomy(
  seqs = seqtab_nochim,
  refFasta = taxo_db,
  multithread = TRUE
)
```

### Getting unqiue sequences

```{r process_data}
dada2::uniquesToFasta(
  dada2::getUniques(seqtab_nochim),
  "uniqueSeqs.fasta",
  ids = as.character(as.list(names(uniques))),
  collapse = FALSE
)


dada2:::writeFasta(dada2::getUniques(seqtab_nochim), "uniqueSeqs.fasta")

# Create a DNAStringSet from the ASVs
dna <- Biostrings::readDNAStringSet(
  filepath = paste0(path_results, "/uniqueSeqs.fasta")
)
```

## Run phyloseq

Create a phyloseq object.

```{r test}
## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 19216 taxa and 8 samples ]
## sample_data() Sample Data:       [ 8 samples by 7 sample variables ]
## tax_table()   Taxonomy Table:    [ 19216 taxa by 7 taxonomic ranks ]
## phy_tree()    Phylogenetic Tree: [ 19216 tips and 19215 internal nodes ]

ps_run <- phyloseq::phyloseq(
  phyloseq::otu_table(seqtab_nochim, taxa_are_rows = FALSE),
  tax_table(taxa),
  phyloseq::refseq(dna))
```

### Visualizations





